### DISCUSSION TITLE
Adding a `Not` operator

### DISCUSSION THREAD

**Discussion by randomeizer** [pointfreeco/swift-parsing#88] [Ideas]
So, in my current context, I'm parsing source code, pulling out documentation comments, which are lines prefixed with `///` (vs `//` or `////`). I'd like to be able to do something like (simplified):

```swift
let docLine = Parse {
    "/// "
    PrefixThrough("\n)
}

let notDocLine = OneOf {
    Not {
        "/// "
    }
    Skip(PrefixThrough("\n"))
}

let docLines = Many {
    Skip {
        Many {
            notDocLine
        }
    }
    docLine
}
```

I'm not quite sure how to build a `Not` operator. Thoughts?

**Comment by randomeizer** [comment]
A related thing I was thinking about is, is it possible to have `PrefixUpTo` and `PrefixThrough` get passed a `Parser` rather than a `String`/`Substring`/etc? Eg:

```swift
let notDocLines = PrefixUpTo {
    "\n"
    oneOrMoreSpaces
    "/// "
}
```

**Reply by stephencelis** [reply]
> is it possible to have `PrefixUpTo` and `PrefixThrough` get passed a `Parser` rather than a `String`/`Substring`/etc?

While it's possible, we're not sure it's a performant thing to do. When working directly with a possible prefix of characters/bytes, we can call down to collection/string methods that are pretty good about scanning the underlying collection for a certain subsequence.

Meanwhile, if we were handed a parser instead, we'd have to run that parser, drop a character, run it again, and keep going. This means running a potentially expensive operation for every character of input till it succeeds (or till we run out of input).

We think it _might_ be possible to have `PrefixUpTo` and `PrefixThrough` detect the more performant parsers (`StartsWith` and the literal parsers) and special case them, but we're worried that the general case is too slow to encourage.

**Reply by randomeizer** [reply]
Yeah, that's definitely a concern (similar to some regex operators), but there are plenty of ways to write a bad parser. In some cases it may the be only way to achieve something.

In most cases, the failure would happen pretty quickly. Would still require testing it by dropping each item in the collection though.

**Reply by stephencelis** [reply]
We're not against it, especially if we can get some benchmarks in place to show that for your average parser it performs OK, but we're worried that it naively introduces performance issues that are quadratic with the input length.

In the meantime it should be implementable outside the library if you want to try it out!

**Reply by randomeizer** [reply]
Yeah, totally, I get the concern. I might have a play.

**Comment by stephencelis** [comment]
> I'm not quite sure how to build a `Not` operator. Thoughts?

I think `Not` would be a great combinator to add! There's even [a precedent](https://docs.rs/nom/latest/nom/combinator/fn.not.html) in [nom](https://github.com/Geal/nom), a popular Rust parsing framework.

I think we could start by getting a parser into place:

```swift
struct Not<Upstream>: Parser where Upstream: Parser {
}
```

We'll need to hold onto that upstream parser:

```swift
let upstream: Upstream
```

And we'll need a builder-friendly initializer:

```swift
init(@ParserBuilder _ build: () -> Upstream) {
  self.upstream = build()
}
```

Now all we need to do is implement `parse` to get a passing conformance, which can call down to the underlying `upstream` parser's `parse` method, and fail if it succeeds (while also rolling back `input` to its original state), or return `Void` to succeed if it fails:

```swift
func parse(_ input: inout Upstream.Input) -> Void? {
  let original = input
  if self.upstream.parse(&input) {
    input = original
    return nil
  }
  return ()
}
```

I've only sketched this out in this GitHub text field, so there may be bugs or compiler errors to work through, but hopefully it's enough to get you started :smile:

**Reply by randomeizer** [reply]
My implementation of `Not` below. Pretty close to your sketch.

```swift
/// Parses a `Void` result if the next input does not match the provided `Upstream` ``Parser``,
/// otherwise returns `nil`, in both cases leaving the input unchanged.
struct Not<Upstream>: Parser where Upstream: Parser {
    let upstream: Upstream
    
    /// Construct a ``Not`` with the provided `Upstream` ``Parser``.
    ///
    /// - Parameter upstream: The ``Parser`` to check.
    init(_ upstream: Upstream) {
        self.upstream = upstream
    }
    
    /// Construct a ``Not`` with the provided `Upstream` ``ParserBuilder`` closure.
    ///
    /// Parameter build: The `Upstream` ``Parser``-returning closure.
    init(@ParserBuilder _ build: () -> Upstream) {
      self.upstream = build()
    }
    
    func parse(_ input: inout Upstream.Input) -> Void? {
      let original = input
      if self.upstream.parse(&input) != nil {
        input = original
        return nil
      }
      return ()
    }
}
```

**Reply by randomeizer** [reply]
And here is an implementation of `Peek`. Not sure if that's the best name for it.

```swift
/// Parses a `Void` result if the next input matches the provided `Upstream` ``Parser``,
/// otherwise returns `nil`, in both cases leaving the input unchanged.
struct Peek<Upstream>: Parser where Upstream: Parser {
    let upstream: Upstream
    
    /// Construct a ``Peek`` with the provided `Upstream` ``Parser``.
    ///
    /// - Parameter upstream: The ``Parser`` to check.
    init(_ upstream: Upstream) {
        self.upstream = upstream
    }
    
    /// Construct a ``Peek`` with the provided `Upstream` ``ParserBuilder`` closure.
    ///
    /// Parameter build: The `Upstream` ``Parser``-returning closure.
    init(@ParserBuilder _ build: () -> Upstream) {
      self.upstream = build()
    }
    
    func parse(_ input: inout Upstream.Input) -> Void? {
      let original = input
      if self.upstream.parse(&input) != nil {
        input = original
        return ()
      }
      return nil
    }
}
```

**Reply by stephencelis** [reply]
@randomeizer Would you be up to PRing those to `main`? Without the `ParserBuilder` stuff (we'll add that when we're closer to release).

So instead your example would be:

```swift
Not("/// ")
  .skip(PrefixThrough("\n"))
```

**Reply by iampatbrown** [reply]
@stephencelis is it worth making these inlinable?

**Reply by stephencelis** [reply]
Yup for sure!

**Reply by randomeizer** [reply]
I'll push a PR shortly.

**Reply by randomeizer** [reply]
PR here: https://github.com/pointfreeco/swift-parsing/pull/89
